bootloader/grub_test.go

```diff
@@ -605,3 +606,159 @@ managed by snapd`))
 	c.Assert(err, ErrorMatches, "cannot load existing config asset: .*/grub.cfg: permission denied")
 	c.Assert(is, Equals, false)
 }
+
+func (s *grubTestSuite) TestRecoveryUpdateBootConfigNoEdition(c *C) {
+	// native EFI/ubuntu setup
+	s.makeFakeGrubEFINativeEnv(c, []byte("recovery boot script"))
+
+	opts := &bootloader.Options{Recovery: true}
+	g := bootloader.NewGrub(s.rootdir, opts)
+	c.Assert(g, NotNil)
+
+	restore := assets.MockInternal("grub-recovery.cfg", []byte(`# Snapd-Boot-Config-Edition: 5
+this is mocked grub-recovery.conf
+`))
+	defer restore()
+
+	eg, ok := g.(bootloader.ManagedAssetsBootloader)
+	c.Assert(ok, Equals, true)
+	// install the recovery boot script
+	err := eg.UpdateBootConfig(opts)
+	c.Assert(err, IsNil)
+
+	c.Assert(filepath.Join(s.grubEFINativeDir(), "grub.cfg"), testutil.FileEquals, `recovery boot script`)
+}
+
+func (s *grubTestSuite) TestRecoveryUpdateBootConfigUpdates(c *C) {
+	// native EFI/ubuntu setup
+	s.makeFakeGrubEFINativeEnv(c, []byte(`# Snapd-Boot-Config-Edition: 1
+recovery boot script`))
+
+	opts := &bootloader.Options{Recovery: true}
+	g := bootloader.NewGrub(s.rootdir, opts)
+	c.Assert(g, NotNil)
+
+	restore := assets.MockInternal("grub-recovery.cfg", []byte(`# Snapd-Boot-Config-Edition: 3
+this is mocked grub-recovery.conf
+`))
+	defer restore()
+	restore = assets.MockInternal("grub.cfg", []byte(`# Snapd-Boot-Config-Edition: 4
+this is mocked grub.conf
+`))
+	defer restore()
+	eg, ok := g.(bootloader.ManagedAssetsBootloader)
+	c.Assert(ok, Equals, true)
+	// install the recovery boot script
+	err := eg.UpdateBootConfig(opts)
+	c.Assert(err, IsNil)
+	// the recovery boot asset was picked
+	c.Assert(filepath.Join(s.grubEFINativeDir(), "grub.cfg"), testutil.FileEquals, `# Snapd-Boot-Config-Edition: 3
+this is mocked grub-recovery.conf
+`)
+}
+
+func (s *grubTestSuite) testBootUpdateBootConfigUpdates(c *C, oldConfig, newConfig string, update bool) {
+	// native EFI/ubuntu setup
+	s.makeFakeGrubEFINativeEnv(c, []byte(oldConfig))
+
+	opts := &bootloader.Options{NoSlashBoot: true}
+	g := bootloader.NewGrub(s.rootdir, opts)
+	c.Assert(g, NotNil)
+
+	restore := assets.MockInternal("grub.cfg", []byte(newConfig))
+	defer restore()
+
+	eg, ok := g.(bootloader.ManagedAssetsBootloader)
+	c.Assert(ok, Equals, true)
+	err := eg.UpdateBootConfig(opts)
+	c.Assert(err, IsNil)
+	if update {
+		c.Assert(filepath.Join(s.grubEFINativeDir(), "grub.cfg"), testutil.FileEquals, newConfig)
+	} else {
+		c.Assert(filepath.Join(s.grubEFINativeDir(), "grub.cfg"), testutil.FileEquals, oldConfig)
+	}
+}
+
+func (s *grubTestSuite) TestNoSlashBootUpdateBootConfigNoUpdateWhenNotManaged(c *C) {
+	oldConfig := `not managed`
+	newConfig := `# Snapd-Boot-Config-Edition: 3
+this update is not applied
+`
+	// the current boot config is not managed, no update applied
+	const updateApplied = false
+	s.testBootUpdateBootConfigUpdates(c, oldConfig, newConfig, updateApplied)
+}
```

Comment:
nitpick, but a lot of these tests would be easier to read / grok if you used a single test with a test table with the updateApplied, oldConfig, and newConfig values in the test case struct, but not a blocker.